### ISSUE TITLE
Cannot enable debug in mpi engines `WARNING | debugpy_stream undefined, debugging will not be enabled`

### ISSUE THREAD

**Issue by lrtfm** [ipython/ipyparallel#639]
When I start the `ipcluster` by
```
ipcluster start --engines=mpi -n 2 --debug
```
The engines report the following warning:
```
2021-11-27 13:31:41.056 [IPEngine.0] WARNING | debugpy_stream undefined, debugging will not be enabled
2021-11-27 13:31:41.061 [IPEngine.1] WARNING | debugpy_stream undefined, debugging will not be enabled
```

How can I enable the debugging in engines? Thanks!

The output of the cluster is attached:
[cluster.log](https://github.com/ipython/ipyparallel/files/7611285/cluster.log)

**Comment by minrk** [comment]
debugpy is not yet available in ipyparallel. I haven't quite figured out how it should work.

Can I ask how were you planning to use the debugpy connections? That would help me figure out how it should work.

**Comment by lrtfm** [comment]
I would like to use [firedrake](https://github.com/firedrakeproject/firedrake), an automated system for the solution of partial differential equations using the finite element method, in ipyparallel. 

There is a function `Function.at` in firedrake. It will compile some c code to a `so` file, load the `so` file, and then call some c function in the `so` file. When I call this function, the engine will stop unexcepted. So I want to know what happened to the engines.

In a jupyter notebook, I start the cluster and engines with
```
import ipyparallel as ipp

cluster = ipp.Cluster(engines="mpi", n=2)
client = cluster.start_and_connect_sync()
```

and run the following code in another cell
```
%%px --block --progress-after=-1

from firedrake import *
from firedrake.petsc import PETSc

mesh = RectangleMesh(10, 10, 1, 1)
V = FunctionSpace(mesh, 'CG', 1)

x, y = SpatialCoordinate(mesh)
u = Function(V).interpolate(sin(x))

PETSc.Sys.Print('Before at')
v = u.at([0.5, 0.5])
PETSc.Sys.Print('End at')
```
Then the code present some errors like this without any more details
```
[stdout:0] Before at
engine set stopped 1638003419: {'exit_code': 1, 'pid': 462554, 'identifier': 'ipengine-1638003418-nbsl-1638003419-462508'}
2 errors
```

**Comment by minrk** [comment]
If the process is crashing, I'm not sure the debugpy debugger will help. That's for interactive debugging of a running session.

Do you get more useful output from:

```python
print(cluster.engine_set.get_output())
```

?

**Comment by lrtfm** [comment]
The output from 
```
print(cluster.engine_set.get_output())
```
is attatched.

[debug.txt](https://github.com/ipython/ipyparallel/files/7617630/debug.txt)

**Comment by minrk** [comment]
Hm, very strange indeed.

Since it's such a small snippet, can you run the same code in a script with `mpiexec -n 2 python script.py`?

If the process is exiting, I would expect some feedback from mpiexec about that.

How have you installed mpi, and which MPI implementation is it?

**Comment by lrtfm** [comment]
The process will be ok when run in a script.
```
(firedrake-complex) z2yang@AMAs3:~/workspace/coding-notes$ mpiexec -n 2 python script.py
Before at
End at
(firedrake-complex) z2yang@AMAs3:~/workspace/coding-notes$
```

The MPI implementation is MPICH 3.4.2.
```
(firedrake-complex) z2yang@AMAs3:~/workspace/coding-notes$ mpiexec --version
HYDRA build details:
    Version:                                 3.4.2
    Release Date:                            Wed May 26 15:51:40 CDT 2021
    CC:                              gcc  -fPIC -fstack-protector -g -O
    Configure options:                       '--disable-option-checking' '--prefix=/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default' 'MAKE=/data/home/z2yang/opt/firedrake-env/.spack-env/view/bin/gmake' '--libdir=/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/lib' 'CC=gcc' 'CFLAGS=-fPIC -fstack-protector -g -O -O2' 'AR=/data/home/z2yang/opt/firedrake-env/.spack-env/view/bin/ar' 'ARFLAGS=cr' 'CXX=g++' 'CXXFLAGS=-fstack-protector -g -O -fPIC -std=gnu++11 -O2' 'FFLAGS=-fPIC -ffree-line-length-0 -g -O -O2' 'FC=gfortran' 'F77=gfortran' 'FCFLAGS=-fPIC -ffree-line-length-0 -g -O -O2' '--enable-shared' '--with-pm=hydra' '--disable-java' '--with-hwloc-prefix=/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default' '--with-device=ch3:nemesis' '--enable-g=meminit' '--cache-file=/dev/null' '--srcdir=.' 'LDFLAGS= -L/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/lib' 'LIBS=' 'CPPFLAGS= -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/src/mpl/include -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/src/mpl/include -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/modules/yaksa/src/frontend/include -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/modules/yaksa/src/frontend/include -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/modules/json-c -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/modules/json-c -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/include -D_REENTRANT -I/data/home/z2yang/opt/firedrake-env/firedrake-complex/src/petsc/default/externalpackages/mpich-3.4.2/src/mpi/romio/include' 'MPLLIBNAME=mpl'
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Demux engines available:                 poll select
```

**Comment by minrk** [comment]
Strange. There shouldn't be anything different between `mpiexec python script.py` and `mpiexec ipengine`. I'm not sure what to try next. Does other MPI code, without firedrake and/or PETSc work?

**Comment by lrtfm** [comment]
Everything will be ok if I don't call the function `at`.

**Comment by minrk** [comment]
Hm. That sounds like I would expect some kind of low-level crash output when you call the function. Very strange that it fails with no feedback.

Can you run the script:


```python
import ipyparallel as ipp

test_code = """
from firedrake import *
from firedrake.petsc import PETSc

mesh = RectangleMesh(10, 10, 1, 1)
V = FunctionSpace(mesh, 'CG', 1)

x, y = SpatialCoordinate(mesh)
u = Function(V).interpolate(sin(x))

PETSc.Sys.Print('Before at')
v = u.at([0.5, 0.5])
PETSc.Sys.Print('End at')
"""

with ipp.Cluster(engines="mpi", n=2, log_level=10) as rc:
    ar = rc[:].execute(test_code, block=False)
    with ar.stream_output():
        ar.get()
```

Maybe that will dump something to the terminal that isn't getting captured.

**Comment by lrtfm** [comment]
Thank you for your patience. It seems that the output doesn't have much useful information.
[cluster.log](https://github.com/ipython/ipyparallel/files/7634726/cluster.log)

**Comment by minrk** [comment]
One more try:

1. start a controller in one terminal (`ipcontroller`)
2. start engines with `mpiexec -verbose -prepend-rank -print-all-exitcodes -n 2 ipengine --debug` in another
3. run the client part of the script, creating the client with `rc = ipp.Client()`

and hope for illuminating output in the engine terminal!

**Comment by lrtfm** [comment]
In this way, the problem is gone. The script and output are shown below. 
Does this give us some hints on the problem before?

```
(firedrake) z2yang@AMAs3:~$ cat client.py
import ipyparallel as ipp

test_code = """
from firedrake import *
from firedrake.petsc import PETSc

mesh = RectangleMesh(10, 10, 1, 1)
V = FunctionSpace(mesh, 'CG', 1)

x, y = SpatialCoordinate(mesh)
u = Function(V).interpolate(sin(x))

PETSc.Sys.Print('At Start')
v = u.at([0.5, 0.5])
PETSc.Sys.Print('value: ', v)
PETSc.Sys.Print('At End')
"""

with ipp.Client() as rc:
    ar = rc[:].execute(test_code, block=False)
    with ar.stream_output():
        ar.get()
(firedrake) z2yang@AMAs3:~$ python client.py
[stdout:1]
At Start
value:  0.47942553860420306
At End
```

**Comment by minrk** [comment]
That should mean that it is something in the [MPIEngineSetLauncher](https://github.com/ipython/ipyparallel/blob/f2674e24e914ce3119d60ca634e59c3f818ed21b/ipyparallel/cluster/launcher.py#L939-L956), which _almost_ only does `mpiexec ipengine --mpi`, but also does some output redirection. Maybe that is the issue.

Does adding `--mpi` after ipengine in the manual example change anything?


I would mostly be poking around pretty blindly, trying to remove things from the Launcher until it works. I don't suppose you have a reproducible environment such as a Dockerfile/image where this error can be reliably triggered?

**Comment by lrtfm** [comment]
I checked the environments in the engines. start the mpi process manually will not have the following environment variables.
Will this be the problem?
```
IPP_CLUSTER_ID = 1638442669-mai4
IPP_PROFILE_DIR = /data/home/z2yang/.ipython/profile_default
IPP_CONNECTION_INFO = {"ssh": "", "interface": "tcp://127.0.0.1", "registration": 59645, "control": 39157, "mux": 37439, "task": 56959, "iopub": 33967, "hb_ping": 54913, "hb_pong": 56409, "broadcast": [52115, 34093], "key": "4af2338f-dbb5e921fc7c2970cdfd38e2", "curve_serverkey": null, "location": "AMAs3", "pack": "json", "unpack": "json", "signature_scheme": "hmac-sha256"}
```

**Comment by lrtfm** [comment]
> That should mean that it is something in the [MPIEngineSetLauncher](https://github.com/ipython/ipyparallel/blob/f2674e24e914ce3119d60ca634e59c3f818ed21b/ipyparallel/cluster/launcher.py#L939-L956), which _almost_ only does `mpiexec ipengine --mpi`, but also does some output redirection. Maybe that is the issue.
> 
> Does adding `--mpi` after ipengine in the manual example change anything?
> 
> I would mostly be poking around pretty blindly, trying to remove things from the Launcher until it works. I don't suppose you have a reproducible environment such as a Dockerfile/image where this error can be reliably triggered?

Add the `--mpi` will result in the problem. see logs below:
```
[proxy:0:0@AMAs3] got pmi command (from 0): abort
exitcode=1
[proxy:0:0@AMAs3] we don't understand this command abort; forwarding upstream
[mpiexec@AMAs3] [pgid: 0] got PMI command: cmd=abort exitcode=1
[proxy:0:0@AMAs3] got pmi command (from 6): abort
exitcode=1
[proxy:0:0@AMAs3] we don't understand this command abort; forwarding upstream
[mpiexec@AMAs3] [pgid: 0] got PMI command: cmd=abort exitcode=1
```

Actually, Firdrake has docker images. I will find the URL and send it to you. Thanks.

**Comment by minrk** [comment]
`--mpi` only adds:

```python
from mpi4py import MPI
mpi_rank = MPI.COMM_WORLD.Get_rank()
mpi_size = MPI.COMM_WORLD.Get_size()
```

Do you still see the same issue if you run without `--mpi` and run that code before your firedrake import? I wonder if `MPI_Init` is getting called twice somehow.

**Comment by lrtfm** [comment]
> 
I can see the same issue if I run without `--mpi` and run that code before my firedrake import.

**Comment by minrk** [comment]
> But why it didn't raise an Exception in the first place when I import the firedrake?

I don't know enough about mpich to know what would be wrong there.

My guess is that firedrake and/or petsc are linked to mpi in 

Just in case, can you add `from firedrake import
Can you add to `ipengine_config.py`:

```python
from firedrake import *
from firedrake.petsc import PETSc
```

Running those very early in the startup may help (before other MPI calls are made). Putting them in the config file itself ensures it will run before any of IPP's own MPI calls.

**Comment by lrtfm** [comment]
Thanks for your help, @minrk. After I import firedrake at the beginning of `ipengine_config.py`, the issue was gone!

Another little problem: how can I make `mpi` the default engine?

Now, I create a profile called mpi, and edit the file ` ~/.ipython/profile_mpi/ipcluster_config.py` with
```
c.Cluster.engine_launcher_class = 'mpi'
```
But, the engine still started with `ipyparallel.cluster.launcher.LocalEngineSetLauncher` other than `ipyparallel.cluster.launcher.MPIEngineSetLauncher`? And this happens when I run the code in a script like this
```
import ipyparallel as ipp

test_code = """
from firedrake import *
from firedrake.petsc import PETSc
import os

mesh = RectangleMesh(10, 10, 1, 1)
mesh.topology_dm.view()
V = FunctionSpace(mesh, 'CG', 1)

x, y = SpatialCoordinate(mesh)
u = Function(V).interpolate(sin(x))

PETSc.Sys.Print('At Start')
v = u.at([0.5, 0.5])
PETSc.Sys.Print('value: ', v)
PETSc.Sys.Print('At End')
"""

with ipp.Cluster(profile="mpi", n=2, log_level=10) as rc:
    ar = rc[:].execute(test_code, block=True)
    with ar.stream_output():
        ar.get()
```
 In Jupyter notebook, the same code will start the `mpi` engines. Any suggestions?

**Comment by lrtfm** [comment]
> I'm not sure i read all that correctly, are you saying that it did load the profile when you ran it in a notebook, but didn't when you ran it as a Python script?

Yes, it seems it didn't load the config file when I ran it as a Python script.

**Comment by minrk** [comment]
Found it, fixed by #643

**Comment by lrtfm** [comment]
Perfect! Thanks!
